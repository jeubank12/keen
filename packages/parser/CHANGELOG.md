# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.2.22](https://github.com/keen/keen/compare/@keen.io/parser@0.2.21...@keen.io/parser@0.2.22) (2020-06-03)

**Note:** Version bump only for package @keen.io/parser





## [0.2.21](https://github.com/keen/keen/compare/@keen.io/parser@0.2.20...@keen.io/parser@0.2.21) (2020-05-28)

**Note:** Version bump only for package @keen.io/parser





## [0.2.20](https://github.com/keen/keen/compare/@keen.io/parser@0.2.19...@keen.io/parser@0.2.20) (2020-05-28)

**Note:** Version bump only for package @keen.io/parser





## [0.2.19](https://github.com/keen/keen/compare/@keen.io/parser@0.2.18...@keen.io/parser@0.2.19) (2020-05-26)

**Note:** Version bump only for package @keen.io/parser





## [0.2.18](https://github.com/keen/keen/compare/@keen.io/parser@0.2.17...@keen.io/parser@0.2.18) (2020-05-22)

**Note:** Version bump only for package @keen.io/parser





## [0.2.17](https://github.com/keen/keen/compare/@keen.io/parser@0.2.16...@keen.io/parser@0.2.17) (2020-05-18)

**Note:** Version bump only for package @keen.io/parser





## [0.2.16](https://github.com/keen/keen/compare/@keen.io/parser@0.2.15...@keen.io/parser@0.2.16) (2020-05-18)

**Note:** Version bump only for package @keen.io/parser





## [0.2.15](https://github.com/keen/keen/compare/@keen.io/parser@0.2.14...@keen.io/parser@0.2.15) (2020-05-15)

**Note:** Version bump only for package @keen.io/parser





## [0.2.14](https://github.com/keen/keen/compare/@keen.io/parser@0.2.13...@keen.io/parser@0.2.14) (2020-05-13)

**Note:** Version bump only for package @keen.io/parser





## [0.2.13](https://github.com/keen/keen/compare/@keen.io/parser@0.2.12...@keen.io/parser@0.2.13) (2020-05-12)

**Note:** Version bump only for package @keen.io/parser





## [0.2.12](https://github.com/keen/keen/compare/@keen.io/parser@0.2.11...@keen.io/parser@0.2.12) (2020-05-11)

**Note:** Version bump only for package @keen.io/parser





## [0.2.11](https://github.com/keen/keen/compare/@keen.io/parser@0.2.10...@keen.io/parser@0.2.11) (2020-05-11)

**Note:** Version bump only for package @keen.io/parser





## [0.2.10](https://github.com/keen/keen/compare/@keen.io/parser@0.2.9...@keen.io/parser@0.2.10) (2020-05-11)

**Note:** Version bump only for package @keen.io/parser





## [0.2.9](https://github.com/keen/keen/compare/@keen.io/parser@0.2.8...@keen.io/parser@0.2.9) (2020-05-08)

**Note:** Version bump only for package @keen.io/parser





## [0.2.8](https://github.com/keen/keen/compare/@keen.io/parser@0.2.7...@keen.io/parser@0.2.8) (2020-05-08)

**Note:** Version bump only for package @keen.io/parser





## [0.2.7](https://github.com/keen/keen/compare/@keen.io/parser@0.2.6...@keen.io/parser@0.2.7) (2020-05-08)

**Note:** Version bump only for package @keen.io/parser





## [0.2.6](https://github.com/keen/keen/compare/@keen.io/parser@0.2.5...@keen.io/parser@0.2.6) (2020-05-07)

**Note:** Version bump only for package @keen.io/parser





## [0.2.5](https://github.com/keen/keen/compare/@keen.io/parser@0.2.4...@keen.io/parser@0.2.5) (2020-05-07)

**Note:** Version bump only for package @keen.io/parser





## [0.2.4](https://github.com/keen/keen/compare/@keen.io/parser@0.2.3...@keen.io/parser@0.2.4) (2020-05-07)

**Note:** Version bump only for package @keen.io/parser





## [0.2.3](https://github.com/keen/keen/compare/@keen.io/parser@0.2.2...@keen.io/parser@0.2.3) (2020-05-04)

**Note:** Version bump only for package @keen.io/parser





## [0.2.2](https://github.com/keen/keen/compare/@keen.io/parser@0.2.1...@keen.io/parser@0.2.2) (2020-04-29)

**Note:** Version bump only for package @keen.io/parser





## [0.2.1](https://github.com/keen/keen/compare/@keen.io/parser@0.2.0...@keen.io/parser@0.2.1) (2020-04-29)

**Note:** Version bump only for package @keen.io/parser





# [0.2.0](https://github.com/keen/keen/compare/@keen.io/parser@0.1.0...@keen.io/parser@0.2.0) (2020-04-27)


### Features

* ðŸŽ¸ table widget ([68f2f2e](https://github.com/keen/keen/commit/68f2f2e55fb99a69aab00a053d283c3d6bf015a4))





# [0.1.0](https://github.com/keen/keen/compare/@keen.io/parser@0.0.2...@keen.io/parser@0.1.0) (2020-04-15)


### Features

* ðŸŽ¸ add support for label mappings ([3e65ee9](https://github.com/keen/keen/commit/3e65ee9fb7429a646c5130cde6058d2e6938ba5f))
* ðŸŽ¸ bubble chart ([6475765](https://github.com/keen/keen/commit/64757650343a4cd81eff6475e109348fd33d27a0))





## 0.0.2 (2020-03-16)

**Note:** Version bump only for package @keen.io/parser
