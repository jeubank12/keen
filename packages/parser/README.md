# `@keen.io/parser`

[![Known Vulnerabilities](https://snyk.io/test/github/keen/keen/badge.svg?targetFile=packages/parser/package.json)](https://snyk.io/test/github/keen/keen?targetFile=packages/parser/package.json)

This package contains set of parsers that are used internally to transform Keen API responses into `@keen.io/charts` friendly format
