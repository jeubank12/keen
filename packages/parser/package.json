{
  "name": "@keen.io/parser",
  "version": "2.1.12",
  "description": "Keen parser",
  "author": "Keen.IO <team@keen.io> (https://keen.io/)",
  "homepage": "https://github.com/keen/keen#readme",
  "license": "MIT",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "module": "src/index.ts",
  "directories": {
    "lib": "dist"
  },
  "files": [
    "dist"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/keen/keen.git"
  },
  "scripts": {
    "prebuild": "rimraf dist && rimraf tsconfig.tsbuildinfo",
    "build": "tsc"
  },
  "bugs": {
    "url": "https://github.com/keen/keen/issues"
  },
  "publishConfig": {
    "access": "public"
  },
  "devDependencies": {
    "@keen.io/ui-core": "^2.5.10",
    "@keen.io/charts": "^2.2.17",
    "@keen.io/charts-utils": "^1.0.2"
  }
}
