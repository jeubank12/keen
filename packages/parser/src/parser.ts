import {
  transformIntervalsFromArray,
  transformAtomicResult,
  transformFromNumber,
} from './utils/transform.utils';

import { AnalysisResult, IntervalResult, AtomicResult, Query } from './types';

import { DEFAULT_VALUE_KEY } from './constants';

export const parseQuery = ({
  result,
}: {
  query: Query;
  result: AnalysisResult;
}) => {
  const keys: Set<string> = new Set();
  const results: Record<string, any>[] = [];

  if (typeof result === 'number') {
    return transformFromNumber(result);
  }

  Array.isArray(result) &&
    result.forEach((partialResult: IntervalResult | AtomicResult) => {
      if ('value' in partialResult && 'timeframe' in partialResult) {
        const { value, timeframe } = partialResult as IntervalResult;

        if (Array.isArray(value)) {
          const { data, keys: dataSetKeys } = transformIntervalsFromArray(
            value
          );
          results.push({ name: timeframe.start, ...data });
          dataSetKeys.forEach(key => keys.add(key));
        }

        if (typeof value === 'number') {
          keys.add(DEFAULT_VALUE_KEY);
          results.push({
            name: timeframe.start,
            [DEFAULT_VALUE_KEY]: value,
          });
        }
      }

      if ('result' in partialResult) {
        const { result, ...properties } = transformAtomicResult(partialResult);
        keys.add(DEFAULT_VALUE_KEY);

        Object.values(properties).forEach(property => {
          results.push({ name: property, value: result });
        });
      }
    });

  return {
    keys: [...keys],
    results,
  };
};
