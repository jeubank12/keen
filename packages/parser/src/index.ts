import { mergeResults } from './merge';
import { parseQuery, parseMultipleQueries } from './parser';
import { mapKeys, createScaleSettings } from './utils';

import { KEEN_VALUE, KEEN_KEY } from './constants';

import { Query, AnalysisResult, Step } from './types';

export {
  parseQuery,
  mapKeys,
  mergeResults,
  parseMultipleQueries,
  createScaleSettings,
  Query,
  Step,
  AnalysisResult,
  KEEN_VALUE,
  KEEN_KEY,
};
