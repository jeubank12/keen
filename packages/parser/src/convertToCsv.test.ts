import { convertToCSV } from './convertToCsv';
import { parseQuery } from './parser';
import {
  countAnalysis,
  countAnalysisWithInterval,
  countAnalysisDoubleGroupByWithInterval,
  countAnalysisGroupByWithInterval,
  countAnalysisDoubleGroupBy,
  funnelAnalysis,
  selectUnique,
} from './api.fixtures';

it('should convert query results to csv format', () => {
  [
    countAnalysis,
    countAnalysisWithInterval,
    countAnalysisDoubleGroupByWithInterval,
    countAnalysisGroupByWithInterval,
    countAnalysisDoubleGroupBy,
    funnelAnalysis,
    selectUnique,
  ].forEach(({ result, steps }) => {
    const parsedQuery = parseQuery({ result, steps });
    const { results } = parsedQuery;
    expect(convertToCSV({ data: results })).toMatchSnapshot();
  });
});
