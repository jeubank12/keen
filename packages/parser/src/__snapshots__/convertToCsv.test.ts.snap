// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@keen.io/parser - convertToCSV() should convert query results to csv format 1`] = `
"keen.key,keen.value
Result,7436
"
`;

exports[`@keen.io/parser - convertToCSV() should convert query results to csv format 2`] = `
"keen.key,keen.value
2019-11-01T00:00:00.000Z,436
2019-12-01T00:00:00.000Z,333
2020-01-01T00:00:00.000Z,2991
2020-02-01T00:00:00.000Z,59
"
`;

exports[`@keen.io/parser - convertToCSV() should convert query results to csv format 3`] = `
"keen.key,Edwidge Danticat Love, Anger, Madness,George R. R. Martin Game of Thrones,J.K. Rowling Harry Potter,Stephen King It,Stephen King The Shining
2020-01-01T00:00:00.000Z,95,719,112,2059,6
2020-02-01T00:00:00.000Z,2,11,1,45,0
"
`;

exports[`@keen.io/parser - convertToCSV() should convert query results to csv format 4`] = `
"keen.key,keen.value
Game of Thrones,730
Harry Potter,113
It,2104
\\"Love, Anger, Madness\\",97
The Shining,6
"
`;

exports[`@keen.io/parser - convertToCSV() should convert query results to csv format 5`] = `
"keen.key,keen.value
\\"Love, Anger, Madness Edwidge Danticat\\",97
Game of Thrones George R. R. Martin,730
Harry Potter J.K. Rowling,113
It Stephen King,2104
The Shining Stephen King,6
"
`;

exports[`@keen.io/parser - convertToCSV() should convert query results to csv format 6`] = `
"keen.key,keen.value
pageviews,1128
signups,317
purchases,89
"
`;

exports[`@keen.io/parser - convertToCSV() should convert query results to csv format 7`] = `
"keen.value
Marketing
IT
null
"
`;
